### ISSUE TITLE
Question:  Justification

### ISSUE THREAD

**Issue by kyprifog** [realestate-com-au/pact#100]
In a situation where both the consumer and provider are under your control, such as a front end application communicating with an API, can you provide even more detailed arguments as to why you would chose to use this approach as opposed to simply allowing the front end application to make real time API calls to a special testing distribution of the API during its integration tests?

**Comment by sawilde** [comment]
For me so I can test that my client can handle the extremely hard to reproduce "error" scenarios in a live environment errors that I may see from talking to System A that may have been caused by its interactions with Systems X, Y & Z.

**Comment by uglyog** [comment]
It also allows me to develop the front-end (or consumer) of the API without the API being there. Also, the tests run as unit tests so are fast. And I can develop a complex micro-service solution without a single integration test.

**Comment by kyprifog** [comment]
@uglyog From an architecture standpoint I can see the obvious benefit of having the front end integration testing suite decoupled from a live version of the API, but Im having trouble getting others on board with it.  I've tried making arguments to the effect that utilizing the live API in integration tests can become brittle from a configuration and availability standpoint but I still haven't hit a homerun with anyone as to why its really worth the extra effort.

**Comment by kyprifog** [comment]
Ive also tried to make the argument that since our API also has an interface with backend components this testing approach can be easily extended to other interfaces as well but I still haven't found a concise way to win over my team to the idea.

**Comment by uglyog** [comment]
It all comes down to confidence and acceptable risk profiles. See http://docs.pact.io/faq/index.html#do-i-still-need-endtoend-tests. Some people can only be convinced by seeing for themselves.

**Comment by NigelThorne** [comment]
Without pact I would test the front end, mocking the services, so that can
easily simulate all valid states and error conditions. I would then also
have tests verifying the services behave as expected.

What this misses is the assurance that all the behaviours and states I
simulates with mocks are possible with the real server. This is the value
that pact adds.

**Comment by kyprifog** [comment]
I'm still not having any luck.  Having any of you went down the road, or seen others go down the road of making live api calls to a special testing distribution as your integration tests?

Is there any case study I can reference to show that this approach can become brittle and hard to maintain?

**Comment by tarciosaraiva** [comment]
Basically all is said above: speed, reliability, confidence. If you are in control of both consumer **and** provider then you probably want to share the Pacts generated between them to ensure changes in one don't break the other. And as you have services calling other services, how do you know their "Pact" is maintained throughout development? Perhaps you will find later in the morning after all the integration tests have run from a commit earlier in the morning. With Pact, you will be enabled to fail fast.

You could of course mock the API calls from the front end as @NigelThorne said. But your confidence is decreased and reliability is shaken. Making live api calls as part of your integration tests will add setup overhead and make everything slower. Your speed is lost and confidence and reliability are arguably better.

**Comment by uglyog** [comment]
I haven't used a special version of the API. But Pact grew out of our frustrations in testing micro-services in an integration test environment. Before Pact we where running all the dependant services in a test environment and then the consumer would make live API calls to the running services.

I found the integration tests were slow, brittle and the feed back loop was long. Errors were difficult to diagnose and 90% of failures were due to environment issues.

The problem you'll encounter is that with a special API, you may not be testing against the actual API that runs in production. The special testing API may hide defects or introduce side effects that don't occur in the actual API.

With Pact, the calls are always made to the actual API provider during the verification phase. In fact, I try to make sure the calls are to the actual candidate that is going to be deployed. So I'm verifying the contract against a release candidate, and that gives me confidence that I can deploy it.

**Comment by mefellows** [comment]
Context is important. 

If you are doing [continuous deployment](https://puppetlabs.com/blog/continuous-delivery-vs-continuous-deployment-whats-diff) (CD) you need a way of automating integration tests between your system and its dependent services. If those services are also doing CD, this challenge is even harder. 

"Which service do I need to spin up, seed with state and test against to be confident in just a few minutes its going to be the same version I hit in Production?". This is the question at the heart of integration testing.

It usually involves pointing at a known 'working' environment that represents Production, or dynamically spinning up real environments to test against. The former is fraught with problems, the main being that now all projects/changes/etc. take a linear path to Production, creating bottlenecks for everyone. Also, they are notoriously not representative of Production in most places I've seen. The latter is expensive in terms of effort, time and cost.

Pact solves that problem, by saying that _it doesn't matter_. All that matters, is that the dependent services can continue to talk the agreed contract before and after a given change. Pact runs against the service that is on it's way to Production, and will fail a build if it can't fulfil its duties as a Provider for its consumers. 

In short:

* Speed: Integration tests are slow, Pact tests are fast
* Reliability: Pact will test that _all_ known consumers will continue to operate after the Provider is deployed to production
* Simplicity: Integration environment configuration, setup and management is hard. Pact doesn't need this
* Cost: Integration environment configuration, setup and management is expensive
* Scalability: If you have to integrate to many services, Pact's approach scales easily. Full E2E environments quickly move towards impossible for anything more than a couple of simple services.

**Comment by kyprifog** [comment]
@tarciosaraiva @uglyog @mefellows.  Thanks for your arguments, hugely helpful, I personally have had some difficulty in verbalizing these points.  I think the main point of contention at this point is the perception that the consumer is simply manually building and maintaining a static JSON document (more or less) with the mocks which could itself become brittle and hard to maintain. 

To expand on this:  The idea is that the mocks are being used in a test suite that reflects a change in the request made will fail, but it is unclear to me what type of tests these are.  For example, https://github.com/realestate-com-au/pact/wiki/Why-Pact-may-not-be-the-best-tool-for-testing-public-APIs seems to rule out the possibility that these are actual end to end tests since you are not testing chains of request/responses, but rather a single request/response in isolation.  Provider states seems to address this to some extent but Im still a bit stuck on situations where the provider actually needs to itself mock out calls to other external services that it needs to make.

Therefore, in the consumer you have some type of unit test whose sole purpose is doing the steps necessary to build up to making an external API call, and then mocking out that call, and allowing the provider to test that asynchronously.  The argument that was made to me is that essentially then the test consists on the consumer side of simply making sure that you are maintaining the JSON document properly, which is completely a manual process.  Something that automatically captures the request and response (similar to https://github.com/vcr/vcr) would seem to make a stronger argument since the mocks don't need to be maintained.  My retort to this argument was that the test has to be robust and actually run real application code leading up to the mock, and then fail if the expected request fails to match to the request that was built by the application (similar to how Webmock works). Your example of building the failing test and then building the mock to then pass that test makes sense to me.

However, in our particular case the consumer is a node application and I can't seem to find any driver for pact in javascript or examples of tests built in node that use the test to illustrate my point more strongly, and illustrate the analogous example in javascript. Right now our API is built with Rails and using committee to construct and validate a json schema we manually build with prmd so we are leaning towards expanding out the responsibility of the schema and generating fixtures based upon it instead to be used in the consumers api tests and then just figuring out some way for the provider to run those tests before deploying to make it reciprocally reliable.

**Comment by bvaughn** [comment]
Hey guys. I'm one of the engineers working with @kyprifog on the JavaScript side. @kyprifog's most recent message touches on our concerns about maintaining JSON files but I wanted to make it a tiny bit more explicit.

@NigelThorne said:
> What this misses is the assurance that all the behaviours and states I simulates with mocks are possible **with the real server**. This is the value that pact adds.

@uglyog said:
> The problem you'll encounter is that with a special API, **you may not be testing against the actual API** that runs in production. The special testing API may hide defects or introduce side effects that don't occur in the actual API.
> With Pact, the calls are always made to the actual API provider during the verification phase. In fact, I try to make sure the calls are to the actual candidate that is going to be deployed. So I'm verifying the contract against a release candidate, and that gives me confidence that I can deploy it.

The above 2 quotes speak to the same concerns I have about the JavaScript side of things. If we are maintaining JSON files then there's no guarantee that those files represent what the _real_ JS client does. (One could change without the other; it relies on human processes to keep them in sync.)

I like many aspects of the Pact system but I feel it is necessary for the generation of the request JSON as well as the verification/assertions about the response JSON to pass through _real_ client code. So far we have only been able to think of ways to generate the _requests_ with our application code (using an approach similar to [DiUS/pact-consumer-js-dsl](https://github.com/DiUS/pact-consumer-js-dsl)).

**Comment by mefellows** [comment]
@bvaughn you and @kyprifog just hit the nail on the head:

>  but I feel it is necessary for the generation of the request JSON as well as the verification/assertions about the response JSON to pass through real client code

and

> The above 2 quotes speak to the same concerns I have about the JavaScript side of things. If we are maintaining JSON files then there's no guarantee that those files represent what the real JS client does. (One could change without the other; it relies on human processes to keep them in sync.)

The Pact file itself should _not_ be maintained manually, it is produced as a side-effect of running http requests through your API Consumer code and not just arbitrary execution of the mock service. Yes, you do need to manually setup these interactions and perhaps there is a future where we could use a [VCR](https://github.com/vcr/vcr) or [Diffy](https://github.com/twitter/diffy) type automation, but today it needs to be setup manually. Automated or not, when something breaks/changes you will likely need 2 humans to get together to discuss why and work out a plan to remediate.

It is in any case irrelevant with respect to _reliability_, because the Provider still has to validate that it responds to each of those agreements, which you have just tested as an integration-style test against your Consumer and shown to be valid insofar as your client code sees it.

**Comment by bvaughn** [comment]
Hey @mefellows. These 2 statements seem to contradict, but I assume I'm just misunderstanding them:

> The Pact file itself should not be maintained manually
> Yes, you do need to manually setup these interactions

I think the part that I'm most concerned about (or understand the least) is the bit about how you "_manually setup these interactions_".  From what I can see- and again I may be missing something obvious- you can programmatically setup the _request_ portion of the interaction (using something like [DiUS/pact-consumer-js-dsl](https://github.com/DiUS/pact-consumer-js-dsl)) but you're stuck writing and maintaining the _response_ portion of the interaction by hand (which seems really error-prone).

**Comment by mefellows** [comment]
I can see why that might be :)

Essentially what I'm saying is that you shouldn't manually edit the pact
file - the '.json' file it spits out. you will still need to set it up in
your test cases. This is the part I was referring to possibly automating
down the track.

Pact is not fool proof, if you put gibberish into the consumer spec you
should expect problems. Once you recognise that integration tests are
difficult and always require humans across API boundaries to
communicate, understanding Pacts' mentality starts to become clear.

Honestly, I would give it a go on a project and iterate on both consumer
and provider. Often by doing things is the best way forward - even if you
decide Pact is not for you.

On Saturday, 26 March 2016, Brian Vaughn <notifications@github.com> wrote:

> Hey @mefellows <https://github.com/mefellows>. These 2 statements seem to
> contradict, but I assume I'm just misunderstanding them:
>
> The Pact file itself should not be maintained manually
> Yes, you do need to manually setup these interactions
>
> I think the part that I'm most concerned about (or understand the least)
> is the bit about how you "*manually setup these interactions*". From what
> I can see- and again I may be missing something obvious- you can
> programmatically setup the *request* portion of the interaction (using
> something like DiUS/pact-consumer-js-dsl
> <https://github.com/DiUS/pact-consumer-js-dsl>) but you're stuck writing
> and maintaining the *response* portion of the interaction by hand (which
> seems really error-prone).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/realestate-com-au/pact/issues/100#issuecomment-201338437>
>


-- 
Matt Fellows
- http://www.onegeek.com.au/
